keps/sig-storage/20191012-storagepool-advplacement.md

```diff
@@ -0,0 +1,364 @@
+* Pod3 has an affinity to AZ2, and its PVCs have an affinity to AZ2-SP1.
+
+If we compare this to the case where we can’t express the PVC affinity to the Storage-Pod, Pod1 and Pod2 may land on the same Storage-Pod and so if that Storage-Pod fails, it would take down two Pods. In the scenario described here, the math will show a higher availability and durability.
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story3-MultiStoragePods](./kep-storagepool-story3.png)
+
+
+### Implementation Details/Notes/Constraints
+
+#### New CRD to surface Storage Pools
+```
+apiVersion: storagepool.storage.k8s.io/v1alpha1
+kind: StoragePool
+metadata:
+  name: storagePool1
+  labels:
+     labelKey: labelValue
+spec:
+  driver: myDriver
+  parameters:
+     csiBackendSpecificOpaqueKey: csiBackendSpecificOpaqueValue
+status:
+  accessibleNodes: 
+  - node1
+  - node2
+  capacity:
+    total: 124676572
+```
+
+#### API definitions for the Storage Pool
+```
+type StoragePool struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+ 
+        // Spec defines the desired state of a storage pool
+        // +optional
+        Spec StorgePoolSpec
+ 
+        // Status represents the current information about a storage pool
+        // +optional
+        Status StoragePoolStatus
+}
+ 
+// StoragePoolSpec describes the attributes of a storage pool
+Type StoragePoolSpec struct {
+    	// Name of the driver
+    	Driver string
+ 
+    	// Opaque parameters describing attributes of the storage pool
+    	+optional
+    	Parameters map[string]string
+}
+ 
+type StoragePoolStatus struct {
+    	// Nodes the storage pool has access to
+    	+optional
+  	AccessibleNodes []string
+  	// Capacity of the storage pool
+  	+optional
+  	Capacity *PoolCapacity
+  	// Last errors happened on the pool
+  	+optional
+  	Errors []StoragePoolError
+}
+ 
+type PoolCapacity struct {
+    	// Total capacity of the storage pool
+    	// +optional
+    	Total *resource.Quantity
+}
+ 
+// Describes an error encountered on the pool
+type StoragePoolError struct {
+    	// Time is the timestamp when the error was encountered.
+    	// +optional
+    	Time *metav1.Time
+ 
+    	// Message details the encountered error
+    	// +optional
+    	Message *string
+}
+```
+
+Note that `StoragePool` are non-namespaced. 
+
+#### CSI changes
+
+* Add a “StoragePoolSelector map[string]string” field in CreateVolumeRequest.
+* Add a ControllerListStoragePools RPC that lists storage pools for CSI drivers that support LIST_STORAGE_POOLS controller capability.
+
+#### New external controller for StoragePool
+
+CSI will be extended with a new ControllerListStoragePools() API, which returns all storage pools available via this CSI driver. A new external K8s controller uses this API upon CSI startup to learn about all storage pools and publish them into the K8s API server using the new StoragePool CRD. 
```

Comment:
Do you mean a new sidecar? Does that new API go into the controller service and thus the new sidecar runs alongside that?

Note that CSI driver's which only support ephemeral inline volumes don't need to implement a controller service. In https://github.com/kubernetes/enhancements/pull/1353 I propose to require that they provide `GetCapacity` and to run `external-provisioner` on each node in a mode where it just takes that information and puts it into the capacity object for the driver.

The decision to extend external-provisioner was based on the observation in the Kubernetes-CSI WG that the proliferation of sidecars is making the release process more complicated and a waste of resources because typically they do get deployed together.

Reply:
This needs to be discussed further.  One solution is to let external-provisioner handle it.  Another solution is to have a separate storage pool sidecar handle it.  The third solution is to follow the snapshot controller split model and have a common storage pool controller and a sidecar.

Reply:
Personally I have no preference and will go with the guidance of the rest of the community.